// @polkadot/typegen@16.5.4/cjs/generate/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateDefaultTx = exports.generateDefaultTsDef = exports.generateDefaultRuntime = exports.generateDefaultRpc = exports.generateDefaultQuery = exports.generateDefaultLookup = exports.generateDefaultInterface = exports.generateDefaultEvents = exports.generateDefaultErrors = exports.generateDefaultConsts = void 0;
var consts_js_1 = require("./consts.js");
Object.defineProperty(exports, "generateDefaultConsts", { enumerable: true, get: function () { return consts_js_1.generateDefaultConsts; } });
var errors_js_1 = require("./errors.js");
Object.defineProperty(exports, "generateDefaultErrors", { enumerable: true, get: function () { return errors_js_1.generateDefaultErrors; } });
var events_js_1 = require("./events.js");
Object.defineProperty(exports, "generateDefaultEvents", { enumerable: true, get: function () { return events_js_1.generateDefaultEvents; } });
var interfaceRegistry_js_1 = require("./interfaceRegistry.js");
Object.defineProperty(exports, "generateDefaultInterface", { enumerable: true, get: function () { return interfaceRegistry_js_1.generateDefaultInterface; } });
var lookup_js_1 = require("./lookup.js");
Object.defineProperty(exports, "generateDefaultLookup", { enumerable: true, get: function () { return lookup_js_1.generateDefaultLookup; } });
var query_js_1 = require("./query.js");
Object.defineProperty(exports, "generateDefaultQuery", { enumerable: true, get: function () { return query_js_1.generateDefaultQuery; } });
var rpc_js_1 = require("./rpc.js");
Object.defineProperty(exports, "generateDefaultRpc", { enumerable: true, get: function () { return rpc_js_1.generateDefaultRpc; } });
var runtime_js_1 = require("./runtime.js");
Object.defineProperty(exports, "generateDefaultRuntime", { enumerable: true, get: function () { return runtime_js_1.generateDefaultRuntime; } });
var tsDef_js_1 = require("./tsDef.js");
Object.defineProperty(exports, "generateDefaultTsDef", { enumerable: true, get: function () { return tsDef_js_1.generateDefaultTsDef; } });
var tx_js_1 = require("./tx.js");
Object.defineProperty(exports, "generateDefaultTx", { enumerable: true, get: function () { return tx_js_1.generateDefaultTx; } });
